# Hacker News Project

The objective of this guided project is to compare two types of posts on Hacker News; Ask HN & Show HN. We will compare the two posts in their frequency and engagement with respect to the time of day, how many points the post receives each hour, etc.

In [1]:
from csv import reader # calling open, reader, & list functions to access dataset
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
list_file = list(read_file)
hn = list_file
print(hn[:4])


[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']]


In [2]:
headers = hn[0] # separating the header row from the dataset 
hn = hn[1:]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [3]:
ask_posts = [] # creating lists to sort dataset by type of post
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower() # making title all lowercase to sort w/o case error
    
    if title.startswith('ask hn'): # startswith function to sort by post title
        ask_posts.append(row)
        
    elif title.startswith('show hn'):
        show_posts.append(row)
        
    else:
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

print(ask_posts[:3])
        
    
    
    

1744
1162
17194
[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14']]


In [4]:
total_ask_comments = 0 # counting all ask post comments
for row in ask_posts:
    ask_comment = int(row[4])
    total_ask_comments += ask_comment
    
avg_ask_comments = total_ask_comments / len(ask_posts) # avg amount of comments on ask posts
template = "Average comments on ask posts are "
print(template + str(avg_ask_comments))
    
total_show_comments = 0 #counting all show post comments
for row in show_posts:
    show_comment = int(row[4])
    total_show_comments += show_comment

avg_show_comments = total_show_comments / len(show_posts) # avg amount of comments on show posts
template = "Average comments on show posts are "
print(template + str(avg_show_comments))
    

    
    

Average comments on ask posts are 14.038417431192661
Average comments on show posts are 10.31669535283993


On average, ask posts are more likely to receive comments compared to show posts by a margin of ~30%. This is logical because an ask post primary objective is to collect responses in the form of a solution/advice to the given question. Show posts invite comments but are posted for the actual contents.

# Ask Posts Engagement Relative to Time

In [5]:
import datetime as dt # import datetime module to work with datetime clauses
result_list = [] # list to populate with post created at & number of comments
for row in ask_posts:
    created_at = row[6]
    num_comments = int(row[4])
    result_list.append([created_at, num_comments])
print(result_list[:4])
    
counts_by_hour = {} # dictionary to create freq table of amount of posts by hour
comments_by_hour = {} # dictionary to create freq table of amount of comments on posts by hour
date_format = '%m/%d/%Y %H:%M' # sample format of datetime objects in dataset 
    
for row in result_list:
    date = row[0]
    comment = row[1]
    date_final = dt.datetime.strptime(date, date_format) # parsing the date using datetime clause
    time = date_final.strftime("%H") # formatting date and isolating the hour using datetime clause
    
    if time in counts_by_hour: # if time value exists in dictionary, adds to post count/ comment tally
        counts_by_hour[time] += 1
        comments_by_hour[time] += comment
        
    else:                      # if new time value, creates a key in dictionary to add post/ comment tally
        counts_by_hour[time] = 1
        comments_by_hour[time] = comment 
        
print(comments_by_hour)
print(counts_by_hour)
        
  
    

[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3]]
{'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}
{'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}


In [6]:
avg_by_hour = [] # list to populate with the hour & average comments per post 
for hour in comments_by_hour:
    comments = comments_by_hour[hour]
    counts = counts_by_hour[hour]
    avg_by_hour.append([hour, (comments/counts)])
    
print(avg_by_hour)

    

[['09', 5.5777777777777775], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['16', 16.796296296296298], ['23', 7.985294117647059], ['12', 9.41095890410959], ['17', 11.46], ['15', 38.5948275862069], ['21', 16.009174311926607], ['20', 21.525], ['02', 23.810344827586206], ['18', 13.20183486238532], ['03', 7.796296296296297], ['05', 10.08695652173913], ['19', 10.8], ['01', 11.383333333333333], ['22', 6.746478873239437], ['08', 10.25], ['04', 7.170212765957447], ['00', 8.127272727272727], ['06', 9.022727272727273], ['07', 7.852941176470588], ['11', 11.051724137931034]]


In [7]:
swap_avg_by_hour = [] # list to swap variable placement of previous list
for row in avg_by_hour:
    hr = row[0] 
    avg = row[1]
    swap_avg_by_hour.append([avg, hr])

print(swap_avg_by_hour)


    
    

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


In [8]:
sorted_swap = sorted(swap_avg_by_hour, reverse = True) # sorted function to order number of comments highest to lowest

In [9]:
string = "Top 5 Hours for Ask Post Comments"
print(string)

for row in sorted_swap[:5]:
    date = dt.datetime.strptime(row[1], "%H")
    time = date.strftime('%H:%M')
    template = "{}: {:.2f} average comments per post" # time & comments rounded to nearest hundreth inserted, respectively
    print(template.format(time, row[0]))
    
    

Top 5 Hours for Ask Post Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


According to our list above, the best time to ask questions is at 3pm with an average of ~39 comments per post. However, the comment threshold appears time sensitive due to the fact that the average comments per post is reduced over 50% in just an hour (4pm). 

# Part Two 
First, we will compare the average number of points between ask and show posts. Then, we will conduct further analysis on the more relevant category.

In [10]:
print(headers)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [11]:
sum_points_ask_posts = 0 # counting argument to aggregate total points in each category
sum_points_show_posts = 0

for post in ask_posts:
    points = int(post[3]) # num_points called from 4th variable in header row
    sum_points_ask_posts += points
    
for post in show_posts:
    points = int(post[3])
    sum_points_show_posts += points
    
print(sum_points_ask_posts/len(ask_posts)) 
print(sum_points_show_posts/len(show_posts))

    

15.061926605504587
27.555077452667813


Since show posts are likely to receive more points on average, we will focus the rest of the analysis there. The objective being to find the optimal time for show posts to accrue points. 

In [12]:
result_show_list = [] # list to populate with each show posts time created and points

for post in show_posts:
    created_at = post[6]
    points = int(post[3])
    result_show_list.append([created_at, points])
print(result_show_list[0:5])   


count_show_posts = {} # dictionaries to make freq table of number of posts and points per post
points_show_posts = {}
date_format_show = '%m/%d/%Y %H:%M'

for row in result_show_list:
    date = row[0]
    points = row[1]
    date_final = dt.datetime.strptime(date, date_format_show)
    time = date_final.strftime('%H')
  
    if time in count_show_posts:
        count_show_posts[time] += 1
        points_show_posts[time] += points
        
    else: 
        count_show_posts[time] = 1
        points_show_posts[time] = points

points_show_posts
count_show_posts 


    

[['11/25/2015 14:03', 26], ['11/29/2015 22:46', 747], ['4/28/2016 18:05', 1], ['7/28/2016 7:11', 3], ['1/9/2016 20:45', 1]]


{'14': 86,
 '22': 46,
 '18': 61,
 '07': 26,
 '20': 60,
 '05': 19,
 '16': 93,
 '19': 55,
 '15': 78,
 '03': 27,
 '17': 93,
 '06': 16,
 '02': 30,
 '13': 99,
 '08': 34,
 '21': 47,
 '04': 26,
 '11': 44,
 '12': 61,
 '23': 36,
 '09': 30,
 '01': 28,
 '10': 36,
 '00': 31}

In [13]:
avg_points_per_hour = [] # list for average points per hour on show posts

for row in points_show_posts:
    points = points_show_posts[row] 
    posts = count_show_posts[row]
    avg_points_per_hour.append([row, points/posts])

avg_points_per_hour

[['14', 25.430232558139537],
 ['22', 40.34782608695652],
 ['18', 36.31147540983606],
 ['07', 19.0],
 ['20', 30.316666666666666],
 ['05', 5.473684210526316],
 ['16', 28.322580645161292],
 ['19', 30.945454545454545],
 ['15', 28.564102564102566],
 ['03', 25.14814814814815],
 ['17', 27.107526881720432],
 ['06', 23.4375],
 ['02', 11.333333333333334],
 ['13', 24.626262626262626],
 ['08', 15.264705882352942],
 ['21', 18.425531914893618],
 ['04', 14.846153846153847],
 ['11', 33.63636363636363],
 ['12', 41.68852459016394],
 ['23', 42.388888888888886],
 ['09', 18.433333333333334],
 ['01', 25.0],
 ['10', 18.916666666666668],
 ['00', 37.83870967741935]]

In [14]:
sorted_show_list = [] # new list to call sorted function on average points in descending order
for row in avg_points_per_hour:
    sorted_show_list.append([row[1],row[0]]) 
    
sorted_show_list = sorted(sorted_show_list,reverse = True)

sorted_show_list

[[42.388888888888886, '23'],
 [41.68852459016394, '12'],
 [40.34782608695652, '22'],
 [37.83870967741935, '00'],
 [36.31147540983606, '18'],
 [33.63636363636363, '11'],
 [30.945454545454545, '19'],
 [30.316666666666666, '20'],
 [28.564102564102566, '15'],
 [28.322580645161292, '16'],
 [27.107526881720432, '17'],
 [25.430232558139537, '14'],
 [25.14814814814815, '03'],
 [25.0, '01'],
 [24.626262626262626, '13'],
 [23.4375, '06'],
 [19.0, '07'],
 [18.916666666666668, '10'],
 [18.433333333333334, '09'],
 [18.425531914893618, '21'],
 [15.264705882352942, '08'],
 [14.846153846153847, '04'],
 [11.333333333333334, '02'],
 [5.473684210526316, '05']]

In [15]:
string_2 = "Top 5 Times for Show Posts Ranked by Points "

print(string_2)

for row in sorted_show_list[0:5]:
    template_2 = 'Show posts at {} receive an average of {:.2f} points'
    date = dt.datetime.strptime(row[1], '%H')
    time = date.strftime('%H:%M')
    
    print(template_2.format(time, row[0]))

Top 5 Times for Show Posts Ranked by Points 
Show posts at 23:00 receive an average of 42.39 points
Show posts at 12:00 receive an average of 41.69 points
Show posts at 22:00 receive an average of 40.35 points
Show posts at 00:00 receive an average of 37.84 points
Show posts at 18:00 receive an average of 36.31 points


Our analysis indicates that show posts receive a higher amount of average points per hour, the best time being 11pm with an average of 42.39 points.

# Part 3 (extra credit work)
In part 3, we will perform similar analysis on our other posts list, focusing on the amount of comments and points, respectively.

In [16]:
total_other_comments = 0 # counting total number of comments on other posts
for row in other_posts:
    other_comment = int(row[4])
    total_other_comments += other_comment
avg_other_comments = total_other_comments/len(other_posts) # calculating average number of comments on other posts
template = "Average comments on other posts are "
print(template + str(avg_other_comments))


Average comments on other posts are 26.8730371059672


In [17]:
other_result_list = [] # list to populate with when post created & number of comments in other posts
for post in other_posts:
    created_at = (post[6])
    comments = int(post[4])
    other_result_list.append([created_at, comments])
    
print(other_result_list[0:5])

    

[['8/4/2016 11:52', 52], ['1/26/2016 19:30', 10], ['6/23/2016 22:20', 1], ['6/17/2016 0:01', 1], ['9/30/2015 4:12', 2]]


In [18]:
counts_by_hour_other = {} # dictionaries to make freq table on number of posts & comments per hour 
comments_by_hour_other = {}
date_format_other = '%m/%d/%Y %H:%M'

for row in other_result_list:
    date = row[0]
    comment = row[1]
    date_final = dt.datetime.strptime(date, date_format_other)
    time = date_final.strftime("%H")
    
    if time in counts_by_hour_other:
        counts_by_hour_other[time] += 1
        comments_by_hour_other[time] += comment
        
    else:
        counts_by_hour_other[time] = 1
        comments_by_hour_other[time] = comment
        
print(comments_by_hour_other)

    
    

{'11': 19532, '19': 26167, '22': 17635, '00': 16544, '04': 10953, '09': 14732, '16': 27959, '18': 29186, '10': 15728, '12': 23944, '20': 21080, '03': 10918, '17': 32727, '14': 30973, '13': 28363, '01': 11536, '23': 16592, '08': 13405, '02': 12254, '21': 20635, '15': 30700, '06': 8714, '05': 9768, '07': 12010}


In [19]:
avg_by_hour_other = [] # list to populate with hour of day & average number of comments per other post
for hour in comments_by_hour_other:
    comments = comments_by_hour_other[hour] 
    counts = counts_by_hour_other[hour]
    avg_by_hour_other.append([hour, (comments/counts)])
    
print(avg_by_hour_other)


[['11', 29.593939393939394], ['19', 26.701020408163266], ['22', 23.265171503957784], ['00', 27.076923076923077], ['04', 24.125550660792953], ['09', 27.588014981273407], ['16', 25.394187102633968], ['18', 26.924354243542435], ['10', 26.612521150592215], ['12', 30.34727503168568], ['20', 23.13940724478595], ['03', 26.825552825552826], ['17', 27.99572284003422], ['14', 32.33089770354906], ['13', 30.896514161220043], ['01', 23.072], ['23', 24.617210682492583], ['08', 27.026209677419356], ['02', 27.786848072562357], ['21', 23.60983981693364], ['15', 29.51923076923077], ['06', 21.357843137254903], ['05', 25.175257731958762], ['07', 26.808035714285715]]


In [20]:
swap_avg_by_hour_other = [] # list to swap previous list variables 
for hour in avg_by_hour_other:
    hr = hour[0]
    avg = hour[1]
    swap_avg_by_hour_other.append([avg,hr])
    
swap_avg_by_hour_other

sorted_swap_other = sorted(swap_avg_by_hour_other, reverse = True) # sorts average number of comments highest to lowest
sorted_swap_other
            

[[32.33089770354906, '14'],
 [30.896514161220043, '13'],
 [30.34727503168568, '12'],
 [29.593939393939394, '11'],
 [29.51923076923077, '15'],
 [27.99572284003422, '17'],
 [27.786848072562357, '02'],
 [27.588014981273407, '09'],
 [27.076923076923077, '00'],
 [27.026209677419356, '08'],
 [26.924354243542435, '18'],
 [26.825552825552826, '03'],
 [26.808035714285715, '07'],
 [26.701020408163266, '19'],
 [26.612521150592215, '10'],
 [25.394187102633968, '16'],
 [25.175257731958762, '05'],
 [24.617210682492583, '23'],
 [24.125550660792953, '04'],
 [23.60983981693364, '21'],
 [23.265171503957784, '22'],
 [23.13940724478595, '20'],
 [23.072, '01'],
 [21.357843137254903, '06']]

In [21]:
string_3 = "Top 5 Times for Other Posts Comments"
print(string_3)

for row in sorted_swap_other[:5]:
    date = dt.datetime.strptime(row[1], '%H')
    time = date.strftime('%H:%M')
    template_3 = "{}: {:.2f} average comments per other post"
    print(template_3.format(time,row[0])) 
    

            

Top 5 Times for Other Posts Comments
14:00: 32.33 average comments per other post
13:00: 30.90 average comments per other post
12:00: 30.35 average comments per other post
11:00: 29.59 average comments per other post
15:00: 29.52 average comments per other post


# Other Posts Ranked by Points

In [22]:
print(headers)



['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [23]:
sum_points_other_posts = 0 # counting all points on other posts

for post in other_posts:
    points = int(post[3])
    sum_points_other_posts += points
    
print(sum_points_other_posts/len(other_posts)) # calculating avg points on other posts 

    
    
    

55.4067698034198


In [24]:
other_points_result_list = [] # list of other posts created & number of points

for post in other_posts:
    points = int(post[3])
    created = post[6]
    other_points_result_list.append([created, points])
print(other_points_result_list[0:5])

count_other_posts = {} # frequency tables for amount of other posts per hour and number of points
points_other_posts = {}
date_format_other = '%m/%d/%Y %H:%M'
for row in other_points_result_list:
    date = row[0]
    points = row[1]
    date_final = dt.datetime.strptime(date, date_format_other)
    time = date_final.strftime('%H')
    if time in count_other_posts:
        count_other_posts[time] += 1
        points_other_posts[time] += points
        
    else:
        count_other_posts[time] = 1
        points_other_posts[time] = points
        
print(points_other_posts)
count_other_posts

[['8/4/2016 11:52', 386], ['1/26/2016 19:30', 39], ['6/23/2016 22:20', 2], ['6/17/2016 0:01', 3], ['9/30/2015 4:12', 8]]
{'11': 37995, '19': 58811, '22': 38079, '00': 35718, '04': 22549, '09': 28802, '16': 59655, '18': 58459, '10': 35746, '12': 45287, '20': 41218, '03': 23167, '17': 67777, '14': 59191, '13': 57398, '01': 25303, '23': 35068, '08': 26830, '02': 25786, '21': 43149, '15': 62964, '06': 18864, '05': 19387, '07': 25461}


{'11': 660,
 '19': 980,
 '22': 758,
 '00': 611,
 '04': 454,
 '09': 534,
 '16': 1101,
 '18': 1084,
 '10': 591,
 '12': 789,
 '20': 911,
 '03': 407,
 '17': 1169,
 '14': 958,
 '13': 918,
 '01': 500,
 '23': 674,
 '08': 496,
 '02': 441,
 '21': 874,
 '15': 1040,
 '06': 408,
 '05': 388,
 '07': 448}

In [25]:
avg_points_hour_other = [] # list containing average points on other posts per hour
for row in points_other_posts: 
    post = count_other_posts[row]
    points = points_other_posts[row]
    avg_points_hour_other.append([row, points/posts])
    
avg_points_hour_other
    

[['11', 1225.6451612903227],
 ['19', 1897.1290322580646],
 ['22', 1228.3548387096773],
 ['00', 1152.1935483870968],
 ['04', 727.3870967741935],
 ['09', 929.0967741935484],
 ['16', 1924.3548387096773],
 ['18', 1885.774193548387],
 ['10', 1153.0967741935483],
 ['12', 1460.8709677419354],
 ['20', 1329.6129032258063],
 ['03', 747.3225806451613],
 ['17', 2186.3548387096776],
 ['14', 1909.3870967741937],
 ['13', 1851.5483870967741],
 ['01', 816.2258064516129],
 ['23', 1131.225806451613],
 ['08', 865.483870967742],
 ['02', 831.8064516129032],
 ['21', 1391.9032258064517],
 ['15', 2031.0967741935483],
 ['06', 608.516129032258],
 ['05', 625.3870967741935],
 ['07', 821.3225806451613]]

In [26]:
sorted_other_list = [] # new list using sorted function to list avg points per hour highest to lowest
for row in avg_points_hour_other:
    sorted_other_list.append([row[1], row[0]])
    
sorted_other_list = sorted(sorted_other_list, reverse = True)

sorted_other_list

[[2186.3548387096776, '17'],
 [2031.0967741935483, '15'],
 [1924.3548387096773, '16'],
 [1909.3870967741937, '14'],
 [1897.1290322580646, '19'],
 [1885.774193548387, '18'],
 [1851.5483870967741, '13'],
 [1460.8709677419354, '12'],
 [1391.9032258064517, '21'],
 [1329.6129032258063, '20'],
 [1228.3548387096773, '22'],
 [1225.6451612903227, '11'],
 [1153.0967741935483, '10'],
 [1152.1935483870968, '00'],
 [1131.225806451613, '23'],
 [929.0967741935484, '09'],
 [865.483870967742, '08'],
 [831.8064516129032, '02'],
 [821.3225806451613, '07'],
 [816.2258064516129, '01'],
 [747.3225806451613, '03'],
 [727.3870967741935, '04'],
 [625.3870967741935, '05'],
 [608.516129032258, '06']]

In [27]:
string_4 = "Top 5 times for Other Posts Ranked by Points"
print(string_4)

for row in sorted_other_list[0:5]:
    template_4 = 'Other posts at {} receive on average {:.2f} points'
    date = dt.datetime.strptime(row[1], '%H')
    time = date.strftime('%H:%M')
    print(template_4.format(time, row[0]))
    
    
    

Top 5 times for Other Posts Ranked by Points
Other posts at 17:00 receive on average 2186.35 points
Other posts at 15:00 receive on average 2031.10 points
Other posts at 16:00 receive on average 1924.35 points
Other posts at 14:00 receive on average 1909.39 points
Other posts at 19:00 receive on average 1897.13 points


Our analysis indicates that the best time for other posts to receive points is at 5pm with an average of 2,186.35 points.